File: rawDMCreader.py
#!/usr/bin/env python3
"""
reads .DMCdata files and displays them
 Michael Hirsch
 GPL v3+ license

NOTE: Observe the dtype=np.int64, this is for Windows Python, that wants to
   default to int32 instead of int64 like everyone else!
    --- we can't use long, because that's only for Python 2.7
 """
from __future__ import division, print_function, absolute_import
from os.path import getsize, expanduser, splitext, isfile
import numpy as np
import argparse
from re import search
from warnings import warn
from six import integer_types
#
try:
    from . import getRawInd as gri #using from another package as submodule
except (ValueError,SystemError):
    import getRawInd as gri #using locally

bpp = 16
nHeadBytes = 4

def goRead(bigfn,xyPix,xyBin,FrameIndReq=None, rawFrameRate=None,startUTC=None,verbose=0):

    bigfn = expanduser(bigfn)
#%% check
    if startUTC:
        raise NotImplementedError('frame UTC timing not yet implemented')
    if rawFrameRate:
        raise NotImplementedError('raw frame rate timing not yet implemented')
#%% setup data parameters
    finf = getDMCparam(bigfn,xyPix,xyBin,FrameIndReq,verbose)
    if finf is None: return None, None

#%% preallocate *** LABVIEW USES ROW-MAJOR ORDERING C ORDER
    data = np.zeros((finf['nframeextract'],finf['supery'],finf['superx']),
                    dtype=np.uint16, order='C')
    rawFrameInd = np.zeros(finf['nframeextract'], dtype=np.int64)

    with open(bigfn, 'rb') as fid:
        for j,i in enumerate(finf['frameind']): #j and i are NOT the same in general when not starting from beginning of file!
            data[j,:,:], rawFrameInd[j] = getDMCframe(fid,i,finf,verbose)

    return data, rawFrameInd,finf
#%% workers
def getserialnum(flist):
    """
    This function assumes the serial number of the camera is in a particular place in the filename.
    Yes, this is a little lame, but it's how the original 2011 image-writing program worked, and I've
    carried over the scheme rather than appending bits to dozens of TB of files.
    """
    sn = []
    for f in flist:
        sn.append(int(search(r'(?<=CamSer)\d{3,6}',f).group()))
    return sn

def animate(i,data,himg,ht):
    #himg = plt.imshow(data[:,:,i]) #slow, use set_data instead
    himg.set_data(data[i,:,:])
    ht.set_text('RelFrame#' + str(i) )
    #'RawFrame#: ' + str(rawFrameInd[jFrm]) +

    draw() #plot won't update without plt.draw()!
    #plt.pause(0.01)
    #plt.show(False) #breaks (won't play)
    return himg,ht

def getDMCparam(bigfn,xyPix,xyBin,FrameIndReq=None,verbose=0):
    bigfn = expanduser(bigfn)
    if not isfile(bigfn): #leave this here, getsize() doesn't fail on directory
        warn('{} is not a file!'.format(bigfn))
        return None

    #np.int64() in case we are fed a float or int
    SuperX = np.int64(xyPix[0]) // np.int64(xyBin[0]) # "//" keeps as integer
    SuperY = np.int64(xyPix[1]) // np.int64(xyBin[1])

    Nmetadata = nHeadBytes//2 #FIXME for DMCdata version 1 only
    PixelsPerImage= SuperX * SuperY
    BytesPerImage = PixelsPerImage*bpp//8
    BytesPerFrame = BytesPerImage + nHeadBytes

# get file size
    fileSizeBytes = getsize(bigfn)

    if fileSizeBytes < BytesPerImage:
        warn('File size {} is smaller than a single image frame!'.format(fileSizeBytes))
        return None


    if fileSizeBytes % BytesPerFrame:
        warn("Looks like I am not reading this file correctly, with BPF: {:d}".format(BytesPerFrame))

    nFrame = fileSizeBytes // BytesPerFrame

    (firstRawInd,lastRawInd) = gri.getRawInd(bigfn,BytesPerImage,nHeadBytes,Nmetadata)
    if verbose > 0:
        print('{} frames, Bytes: {} in file {}'.format(nFrame,fileSizeBytes,bigfn))
        print("first / last raw frame #'s: {}  / {} ".format(firstRawInd,lastRawInd))

# setup frame indices
# if no requested frames were specified, read all frames. Otherwise, just
# return the requested frames
    #note these assignments have to be "long", not just python "int", because on windows python 2.7 64-bit on files >2.1GB, the bytes will wrap
    if isinstance(FrameIndReq,integer_types): #the user is specifying a step size
        FrameInd =np.arange(0,nFrame,FrameIndReq,dtype=np.int64)
    elif FrameIndReq and len(FrameIndReq) == 3: #catch is None
        FrameInd =np.arange(FrameIndReq[0],FrameIndReq[1],FrameIndReq[2],dtype=np.int64)
    else: #catch all
        FrameInd = np.arange(nFrame,dtype=np.int64) # has to be numpy.arange for > comparison
        if verbose>0:
            print('automatically selected all frames in file')
    badReqInd = (FrameInd>nFrame) | (FrameInd<0)
# check if we requested frames beyond what the BigFN contains
    if badReqInd.any():
        warn('You have requested Frames ' + str(FrameInd[badReqInd]) +
                 ', which exceeds the length of {}'.format(bigfn))
        return None

    nFrameExtract = FrameInd.size #to preallocate properly

    nBytesExtract = nFrameExtract * BytesPerFrame
    if verbose > 0:
        print('Extracted {} frames from {} totaling {} bytes.'.format(
                   nFrameExtract,bigfn,nBytesExtract))
    if nBytesExtract > 4e9:
        warn('This will require {:.2f} Gigabytes of RAM.'.format(nBytesExtract/1e9))

    return {'superx':SuperX, 'supery':SuperY, 'nmetadata':Nmetadata,
            'bytesperframe':BytesPerFrame, 'pixelsperimage':PixelsPerImage,
            'nframe':nFrame, 'nframeextract':nFrameExtract,'frameind':FrameInd}

def getDMCframe(f,iFrm,finf,verbose=0):
    # on windows, "int" is int32 and overflows at 2.1GB!  We need np.int64
    currByte = iFrm * finf['bytesperframe']
#%% advance to start of frame in bytes
    if verbose>0:
        print('seeking to byte ' + str(currByte))

    assert currByte.dtype == np.int64
    try:
        f.seek(currByte,0) #no return value
    except IOError as e:
        warn('I couldnt seek to byte {:d}'.format(currByte))
        print('try using a 64-bit integer for iFrm')
        print('is ' + str(f.name) +' a valid .DMCdata file?')
        print(str(e))
        return None, None
#%% read data ***LABVIEW USES ROW-MAJOR C ORDERING!!
    try:
        currFrame = np.fromfile(f, np.uint16,
                            finf['pixelsperimage']).reshape((finf['supery'],finf['superx']),
                            order='C')
    except ValueError as e:
        warn('we may have read past end of file?  {}'.format(e))
        return None,None

    rawFrameInd = gri.meta2rawInd(f,finf['nmetadata'])
    return currFrame,rawFrameInd

def doPlayMovie(data,finf,playMovie,Clim,rawFrameInd):
    if playMovie:
        sfmt = ScalarFormatter(useMathText=True)
        hf1 = figure(1)
        hAx = hf1.gca()
        if Clim is None:
            hIm = hAx.imshow(data[0,:,:], cmap = 'gray', origin='lower',norm=LogNorm() )
        else:
            hIm = hAx.imshow(data[0,:,:],
                            vmin=Clim[0],vmax=Clim[1],
                            cmap = 'gray', origin='lower', norm=LogNorm())
        hT = hAx.text(0.5,1.005,'', ha='center',transform=hAx.transAxes)
        hc = hf1.colorbar(hIm,format=sfmt)
        hc.set_label('data numbers ' + str(data.dtype))
        hAx.set_xlabel('x-pixels')
        hAx.set_ylabel('y-pixels')

        for j,i in enumerate(rawFrameInd):
            #print('raw {}  rel {} '.format(i,j))
            #hAx.imshow(data[j,...]) #slower
            hIm.set_data(data[j,...])  # faster
            hT.set_text('RawFrame#: {} RelFrame# {}'.format(rawFrameInd[j],i) )
            draw(); pause(playMovie)

def doanimate(data,nFrameExtract,playMovie):
    # on some systems, just freezes at first frame
    print('attempting animation')
    fg = figure()
    ax = fg.gca()
    himg = ax.imshow(data[:,:,0],cmap='gray')
    ht = ax.set_title('')
    fg.colorbar(himg)
    ax.set_xlabel('x')
    ax.set_ylabel('y')

    #blit=False so that Title updates!
    anim.FuncAnimation(fg,animate,range(nFrameExtract),fargs=(data,himg,ht),
                       interval=playMovie, blit=False, repeat_delay=1000)

def doplotsave(bigfn,data,rawind,clim,dohist,meanImg,writeFITS,saveMat):
    outStem = splitext(expanduser(bigfn))[0]

    if dohist:
        ax=figure().gca()
        hist(data.ravel(), bins=256,log=True)
        ax.set_title('histogram of {}'.format(bigfn))
        ax.set_ylabel('frequency of occurence')
        ax.set_xlabel('data value')

    if meanImg:
        meanStack = data.mean(axis=0).astype(np.uint16) #DO NOT use dtype= here, it messes up internal calculation!
        fg = figure(32)
        ax = fg.gca()
        if clim:
            hi=ax.imshow(meanStack,cmap='gray',origin='lower', vmin=clim[0], vmax=clim[1],norm=LogNorm())
        else:
            hi=ax.imshow(meanStack,cmap='gray',origin='lower',norm=LogNorm())

        ax.set_xlabel('x')
        ax.set_ylabel('y')
        ax.set_title('mean of image frames')
        fg.colorbar(hi)

        pngfn = outStem + '_mean.png'
        print('writing mean PNG ' + pngfn)
        fg.savefig(pngfn,dpi=150,bbox_inches='tight')
#%% saving
    if writeFITS:
        from astropy.io import fits
        #TODO timestamp frames
        if meanImg:
            fitsFN = outStem + '_mean_frames.fits'
            fitsData = meanStack
            print('writing MEAN image data as ' + fitsFN)
        else:
            fitsFN = outStem + '_frames.fits'
            fitsData = rawImgData
            print('writing ' + str(fitsData.dtype) + ' raw image data as ' + fitsFN)
        hdu = fits.PrimaryHDU(fitsData)
        hdu.writeto(fitsFN,clobber=True)
        print('the orientation of this FITS in NASA FV program and the preview image shown in Python should/must have the same orientation and pixel indexing')


    if saveMat:
        from scipy.io import savemat
        matFN = outStem + '_frames.mat'
        print('writing raw image data as ' + matFN)
        matdata = {'rawimgdata':rawImgData}
        savemat(matFN,matdata,oned_as='column')


if __name__ == "__main__":
    from matplotlib.pyplot import figure,show, hist, draw, pause
    from matplotlib.colors import LogNorm
    from matplotlib.ticker import ScalarFormatter
    #import matplotlib.animation as anim
    p = argparse.ArgumentParser(description='Raw .DMCdata file reader')
    p.add_argument('infile',help='.DMCdata file name and path',type=str,nargs='?',default='')
    p.add_argument('-p','--pix',help='nx ny  number of x and y pixels respectively',nargs=2,default=(512,512),type=int)
    p.add_argument('-b','--bin',help='nx ny  number of x and y binning respectively',nargs=2,default=(1,1),type=int)
    p.add_argument('-f','--frames',help='frame indices of file (not raw)',nargs=3,metavar=('start','stop','stride'), type=np.int64) #don't use string
    p.add_argument('-m','--movie',help='seconds per frame. ',type=float)
    p.add_argument('-c','--clim',help='min max   values of intensity expected (for contrast scaling)',nargs=2,type=float)
    p.add_argument('-r','--rate',help='raw frame rate of camera',type=float)
    p.add_argument('-s','--startutc',help='utc time of nights recording')
    p.add_argument('--fits',help='write a .FITS file of the data you extract',action='store_true')
    p.add_argument('--mat',help="write a .mat MATLAB data file of the extracted data",action='store_true')
    p.add_argument('--avg',help='return the average of the requested frames, as a single image',action='store_true')
    p.add_argument('--hist',help='makes a histogram of all data frames',action='store_true')
    p.add_argument('-v','--verbose',help='debugging',action='count',default=0)
    a = p.parse_args()

    rawImgData,rawInd,finf = goRead(a.infile, a.pix, a.bin,a.frames, a.rate,a.startutc,a.verbose)
#%% plots and save
    doPlayMovie(rawImgData,finf,a.movie, a.clim,rawInd)

    doplotsave(a.infile,rawImgData,rawInd,a.clim,a.hist,a.avg,a.fits,a.mat)
    show()

File: rawDMCreader.m
% [data, rawFrameInd,tUTC] = rawDMCreader(BigFN,xPix,yPix,xBin,yBin,FrameInd,playMovie,Clim,rawFrameRate,startUTC)
% 
% reads uint16 raw data files from DMC
% Tested with Octave 3.6, 3.8 & Matlab R2014a
% Michael Hirsch Dec 2011 / Mar 2012 / Mar 2014
%
% requires: getRawInd.m (custom written script)
%
% INPUTS:
% BigFN: huge .DMCdata filename to read
% xPix: # of x-pixels in sensor {512}
% yPix: # of y-pixels in sensor {512}
% xBin: horizontal binning factor of CCD {1}
% yBin: vertical binning factor of CCD {1}
% FrameInd: Frame numbers to extract from file -- this is NOT raw frame numbers,
% but rather 1-indexed from start of file
% playMovie: if ~= 0, play with pause moviePlay seconds
% clim: for imagesc, sets the upper and lower contrast. E.g. [1000, 1200]
% rawFrameRate: {'auto'} get from XML file (recommended) or manually specify 
% startUTC: {'auto'} get from NMEA file (recommended) or manually specify as  "datenum"
%
% OUTPUTS:
% data: 16-bit data, sorted into frames (view with imagesc)
% rawFrameInd: camera index since acquisition start (used to obtain UTC time of
% frame based on GPS)
% tUTC: estimated UTC time of frame -- unverified
%
% Examples:
% [data,~,tUT] = rawDMCreader('~/HSTdata/DataField/2013-04-14/HST1/2013-04-14T07-00-CamSer7196_frames_363000-1-369200.DMCdata',512,512,1,1,1:100,0.01,[100,4000],'auto','auto');
%
% meteor example  (playing first 100 frames of video):
% [data,~,tUT] = rawDMCreader('/cygdrive/d/2014-03-30/2014-03-30T10-46-CamSer7196.DMCdata',512,512,1,1,1:100,0.01,[100,2000],'auto','auto');
%
% meteor example (extracting first 1000 frames of video, then saving to .mat):
% [data,~,tUT] = rawDMCreader('/cygdrive/d/2014-03-30/2014-03-30T10-46-CamSer7196.DMCdata',512,512,1,1,1:1000,0,[],'auto','auto');
% save('first100.mat','-V7')
% of course one could use fitswrite, h5write, etc. for whatever format one wants.
%
%------------
% flintmax('double') is 2^53 (biggest integer value we can hold in a
% double float variable in Matlab on a 64-bit Operating System)

function [data, rawFrameInd, tUTC] =...
       rawDMCreader(BigFN,xPix,yPix,xBin,yBin,FrameInd,playMovie,Clim,rawFrameRate,startUTC,verbose)

if nargin<1, error('you must specify a file to read'), end
if nargin<2, xPix = 512, yPix = 512, end %#ok<NOPRT> %pixels
if nargin<4, xBin = 1, yBin = 1, end %#ok<NOPRT>
if nargin<6, FrameInd = 'all'; end
if nargin<7, playMovie = 0; end
if nargin<8, Clim = []; end 
if nargin<9, rawFrameRate=[]; end
if nargin<10 || isempty(rawFrameRate), startUTC=[]; end
if nargin<11, verbose = false; end
%%
[rawFrameRate,startUTC] = DMCtimeparams(BigFN,rawFrameRate,startUTC);
%% setup data parameters
% based on settings from .xml file (these stay pretty constant)
SuperX = xPix/xBin;
SuperY = yPix/yBin;
bpp = 16; %bits per pixel
nHeadBytes = 4; %bytes per header frame (32 bits for CCD .DMCdata)
nHeader = nHeadBytes/2; % one 16-bit word = 2 bytes
dFormat = 'uint16=>uint16';  %for Andor CCD
PixelsPerImage= SuperX * SuperY;
BytesPerImage = PixelsPerImage*bpp/8;
BytesPerFrame = BytesPerImage + nHeadBytes;

% get file size
fileInfo= dir(BigFN); 
if isempty(fileInfo), error(['file does not exist: ',BigFN]), end
fileSizeBytes = fileInfo.bytes;

if fileSizeBytes < BytesPerImage
    error(['File size ',int2str(fileSizeBytes),' is smaller than a single image frame!'])
end

nFrame = fileSizeBytes / BytesPerFrame; % by inspection
 %there should be no partial frames

if rem(nFrame,1) ~= 0
    warning(['Looks like I am not reading this file correctly, with BPF ',int2str(BytesPerFrame)])
end

%% get "raw" frame numbers -- that Camera FPGA tags each frame with 
% this raw frame is critical to knowing what time an image was taken, which
% is critical for the usability of this data in light of other sensors
% (radar, optical)
[firstRawInd, lastRawInd] = getRawInd(BigFN,BytesPerImage,nHeadBytes);
if verbose
    display([int2str(nFrame),' frames in file ',BigFN])
    display(['   file size in Bytes: ',sprintf('%ld',fileSizeBytes)])
    display(['first / last raw frame # ',int2str(firstRawInd),' / ',...
             int2str(lastRawInd)])
end
% if no requested frames were specified, read all frames. Otherwise, just
% return the requested frames
if strcmpi(FrameInd,'all')
    FrameInd = 1:nFrame;
end
badReqInd = FrameInd > nFrame;
% check if we requested frames beyond what the BigFN contains
if any(badReqInd)
    error(['You have requested Frames ',int2str(FrameInd(badReqInd)),', which exceed the length of the BigFN'])
end
nFrameExtract = length(FrameInd); %to preallocate properly

nBytesExtract = nFrameExtract*BytesPerFrame;
if verbose
    display(['Extracting ',sprintf('%ld',nBytesExtract),' bytes.'])
end
if nBytesExtract > 2e9
    warning(['This will require ',num2str(nBytesExtract/1e9,'%0.1f'),...
            ' Gigabytes of RAM. Do you have enough RAM?'])
end
%% preallocate
% note: Matlab's default data type is "double float", which takes 64-bits
% per number. That can use up all the RAM of your PC. The data is only
% 16-bit, so to load bigger files, I keep the data 16-bit.
% In analysis programs, we might convert the data frame-by-frame to double
% or single float as we stream the data through an algorithm.  
% That's because many Matlab functions will crash or give unexpected
% results with integers (or single float!)
data = zeros(SuperX,SuperY,nFrameExtract,'uint16'); 
% I created a custom header, but I needed 32-bit numbers. So I stick two
% 16-bit numbers together when writing the data--Matlab needs to unstick
% and restick this number into a 32-bit integer again.
% then I store as double in case we want to do numerical operations --
% uint's can lead to unexpected results!
rawFrameInd = zeros(nFrameExtract,1,'double');
if ~isempty(rawFrameRate)
    tUTC = nan(nFrameExtract,1,'double'); 
else
    tUTC = [];
end
%% read data
fid = fopen(BigFN,'r');
if fid<1, error(['error opening ',BigFN]), end
jFrm = 0;
for iFrame = FrameInd  
    
    jFrm = jFrm + 1; %used for indexing memory
    
    currByte = (iFrame - 1) * BytesPerFrame; %start at beginning of frame
    
    % advance to start of frame in bytes
    fsErr = fseek(fid,currByte,'bof');
    if fsErr
        error(['Could not seek to byte ',currByte,', request ',int2str(jFrm)])
    end
    
    %read data
    %we transpose because Labview writes ROW MAJOR and Matlab is COLUMN MAJOR
    data(:,:,jFrm) = transpose(fread(fid,[SuperX,SuperY],dFormat,0,'l')); %first read the image
    metadata = fread(fid,nHeader,dFormat,0,'l'); % we have to typecast this
    
    %stick two 16-bit numbers together again to make the actual 32-bit raw
    %frame index
    rawFrameInd(jFrm) = double( typecast( [metadata(2), metadata(1)] ,'uint32') );
    if ~isempty(rawFrameRate)
        tUTC(jFrm) = startUTC + ( rawFrameInd(jFrm) - 1 )/rawFrameRate /86400; 
    end
end

fclose(fid);
%% play movie, if user chooses
if playMovie   % ~= 0
  doPlayMovie(data,SuperX,SuperY,nFrameExtract,playMovie,Clim,rawFrameInd,tUTC)
end %for
%% cleanup
if ~nargout, clear, end
end %function

function doPlayMovie(data,nRow,nCol,nFrameExtract,playMovie,Clim,rawFrameInd,tUTC)
%% setup plot
% note: this will plot slowly in Octave 3.6, but Octave 3.8 with FLTK will
% plot this about as fast as Matlab
    h.f = figure(1); clf(1)
    h.ax = axes('parent',h.f);
    switch isempty(Clim)
        case false, h.im = imagesc(zeros(nRow,nCol,'uint16'),Clim);
        case true,  h.im = imagesc(zeros(nRow,nCol,'uint16'));
    end %switch
   %flip the picture back upright.  Different programs have different ideas
   %about what corner of the image is the origin. Or whether to start indexing at (0,0) or (1,1).
    set(h.ax,'ydir','normal') 
    % just some labels
    h.t = title(h.ax,'');
    try
        colormap(h.ax,'gray') %matlab, octave 3.8
    catch
        colormap('gray') %octave 3.6
    end
    h.cb = colorbar('peer',h.ax);
    ylabel(h.cb,'Data Numbers')
    xlabel(h.ax,'x-pixels')
    ylabel(h.ax,'y-pixels')
%% do the plotting    
% setting Cdata like this is much, MUCH faster than repeatedly calling
% imagesc() !
try
    for iFrame = 1:nFrameExtract
     set(h.im,'cdata',single(data(:,:,iFrame))) %convert to single just as displaying
     titlestring = ['Raw Frame # ',int2str(rawFrameInd(iFrame)),...
           '  Relative Frame # ',int2str(iFrame)];
     if ~isempty(tUTC)
         titlestring = [titlestring,'  time: ',datestr(tUTC(iFrame),'yyyy-mm-ddTHH:MM:SS.FFF'),' UTC']; %#ok<AGROW>
     end
     
     set(h.t,'String',titlestring)
     pause(playMovie)
    end
catch ME
    display(iFrame)
    try
    display(tUTC(iFrame))
    display(rawFrameInd(iFrame))
    end
    rethrow(ME)
end
%% cleanup
if nargout==0, clear, end
end %function
